cudamapper/src/main.cu

```diff
@@ -397,415 +182,341 @@ void align_overlaps(DefaultDeviceAllocator allocator,
 
-    if (parameters.all_to_all)
+    // load indices into host memory
     {
-        target_parser = query_parser;
+        assert(!host_batch.query_indices.empty() && !host_batch.target_indices.empty() && !device_batches.empty());
+
+        CGA_NVTX_RANGE(profiler, "main::process_one_batch::host_indices");
+        host_cache.generate_query_cache_content(host_batch.query_indices,
+                                                device_batches.front().query_indices,
+                                                skip_copy_to_host);
+        host_cache.generate_target_cache_content(host_batch.target_indices,
+                                                 device_batches.front().target_indices,
+                                                 skip_copy_to_host);
     }
-    else
-    {
-        target_parser = io::create_kseq_fasta_parser(parameters.target_filepath, parameters.k + parameters.w - 1);
-    }
-
-    std::cerr << "Query file: " << parameters.query_filepath << ", number of reads: " << query_parser->get_num_seqences() << std::endl;
-    std::cerr << "Target file: " << parameters.target_filepath << ", number of reads: " << target_parser->get_num_seqences() << std::endl;
-
-    // Data structure for holding overlaps to be written out
-    std::mutex overlaps_writer_mtx;
 
-    struct QueryTargetsRange
+    // process device batches one by one
+    for (const IndexBatch& device_batch : batch.device_batches)
     {
-        IndexDescriptor query_range;
-        std::vector<IndexDescriptor> target_ranges;
-    };
-
-    ///Factor of 1000000 to make max cache size in MB
-    std::vector<IndexDescriptor> query_index_descriptors  = group_reads_into_indices(*query_parser,
-                                                                                    parameters.index_size * 1000000);
-    std::vector<IndexDescriptor> target_index_descriptors = group_reads_into_indices(*target_parser,
-                                                                                     parameters.target_index_size * 1000000);
+        process_one_device_batch(device_batch,
+                                 device_cache,
+                                 application_parameters,
+                                 device_allocator,
+                                 overlaps_and_cigars_to_write,
+                                 cuda_stream);
+    }
+}
 
-    //First generate all the ranges independently, then loop over them.
-    std::vector<QueryTargetsRange> query_target_ranges;
+/// \brief does post-processing and writes data to output
+/// \param device_id
+/// \param application_parameters
+/// \param overlaps_and_cigars_to_write new data is added as it gets available
+/// \param output_mutex controls access to output to prevent race conditions
+void writer_thread_function(const int32_t device_id,
```

Comment:
I find the name a bit misleading, because this function also does post-processing. How about renaming it to `postprocess_and_write_thread_function` or similar.

Reply:
done